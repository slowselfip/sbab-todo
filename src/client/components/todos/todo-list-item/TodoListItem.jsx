import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import './TodoListItem.less';
import TodoCheckbox from '../../checkbox/TodoCheckbox';

const TodoListItem = React.memo(({ todo: { id, completed, description }, onClick, onDelete }) => {
  const handleTodoClick = useCallback(() => onClick(id, completed), [id, completed, onClick]);
  const handleRemoveClick = useCallback(
    event => {
      event.stopPropagation();
      onDelete(id);
    },
    [id, onDelete]
  );
  return (
    <li>
      <div
        role="button"
        tabIndex={0}
        className={cx('todo', { 'todo--completed': completed })}
        onClick={handleTodoClick}
      >
        <div className="todo__checkbox-wrapper">
          <TodoCheckbox className="todo__checkbox" id={id} checked={completed} />
        </div>
        <div className="todo__description">{description}</div>
        <div className="todo__actions">
          <FontAwesomeIcon
            role="button"
            className="todo__action-icon"
            icon="trash"
            onClick={handleRemoveClick}
          />
        </div>
      </div>
    </li>
  );
});

TodoListItem.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.string,
    completed: PropTypes.bool,
    description: PropTypes.string,
    createdAt: PropTypes.string,
    updatedAt: PropTypes.string
  }).isRequired,
  onClick: PropTypes.func,
  onDelete: PropTypes.func
};

TodoListItem.defaultProps = {
  onClick: undefined,
  onDelete: undefined
};

export default TodoListItem;
