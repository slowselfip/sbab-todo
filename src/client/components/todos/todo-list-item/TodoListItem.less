@import (reference) "../../../styles/variables.less";

.todo {
  font-size: 1.25rem;
  display: flex;
  padding: 1rem;
  border: 1px solid;
  margin-bottom: 3px;
  transition: background-color 0.2s ease;
  cursor: pointer;

  &__checkbox-wrapper {
    display: flex;
    align-items: center;
    font-size: 0.5em;
    > label {
      cursor: pointer
    }
  }
  &__description {
    display: flex;
    flex: 1;
    align-items: center;
    margin-left: 1rem;
  }
  &__checkbox {
    cursor: pointer;
  }
  &__actions {
    display: flex;
    align-items: center;
    padding: 0.25rem;
  }
  &__action-icon {
    font-size: 1rem;
  }
  &--completed {
    .todo__description {
      text-decoration: line-through;
    }
  }
}

.light {
  @import (reference) "../../../styles/light";
  .todo {
    background: @top-layer-bg;
    border-color: @border-color;
    &:hover {
      background-color: @top-layer-highlight
    }
    &__action-icon {
      color: @icon-color;
      &:hover {
        color: @icon-color-hover;
      }
    }
  }
}

.dark {
  @import (reference) "../../../styles/dark";
  .todo {
    background: @top-layer-bg;
    border-color: @border-color;
    &:hover {
      background-color: @top-layer-highlight
    }
    &__action-icon {
      color: @icon-color;
      &:hover {
        color: @icon-color-hover;
      }
    }
  }
}