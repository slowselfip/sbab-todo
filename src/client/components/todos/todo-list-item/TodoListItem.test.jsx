import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faTrash } from '@fortawesome/free-solid-svg-icons';
import TodoListItem from './TodoListItem';

library.add(faTrash);

const todo = {
  id: '5cdb2989f7a1683bca97aa02',
  completed: false,
  description: 'Mock todo',
  createdAt: '2019-05-14T20:48:09.189Z',
  updatedAt: '2019-05-14T20:48:09.189Z'
};

afterEach(cleanup);

describe('TodoListItem', () => {
  it('renders', () => {
    const { asFragment } = render(<TodoListItem todo={todo} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Call the onClick callback when clicked', () => {
    const onClickSpy = jest.fn();
    const { container } = render(<TodoListItem todo={todo} onClick={onClickSpy} />);
    const todoItem = container.querySelector(`div[role="button"]`);
    fireEvent.click(todoItem);

    expect(onClickSpy).toHaveBeenCalledWith(todo.id, todo.completed);
  });

  it('Call the onDelete callback when clicking on the trashcan', () => {
    const onDeleteSpy = jest.fn();
    const { container } = render(<TodoListItem todo={todo} onDelete={onDeleteSpy} />);
    const trashcan = container.querySelector(`.todo__action-icon[role="button"]`);
    fireEvent.click(trashcan);

    expect(onDeleteSpy).toBeCalledWith(todo.id);
  });
});
