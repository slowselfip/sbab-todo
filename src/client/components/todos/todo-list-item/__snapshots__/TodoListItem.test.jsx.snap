// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem renders 1`] = `
<DocumentFragment>
  <li>
    <div
      class="todo"
      role="button"
      tabindex="0"
    >
      <div
        class="todo__checkbox-wrapper"
      >
        <label
          for="5cdb2989f7a1683bca97aa02"
        >
          <input
            class="todo-checkbox todo__checkbox"
            id="5cdb2989f7a1683bca97aa02"
            readonly=""
            type="checkbox"
          />
        </label>
      </div>
      <div
        class="todo__description"
      >
        Mock todo
      </div>
      <div
        class="todo__actions"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash fa-w-14 todo__action-icon"
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="button"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
